### PR TITLE
feat: implement promotion resources

### PR DESCRIPTION
**Pull request by Iloniavo**

### CODE REVIEW

File: src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java [hunk 1]
```diff
@@ -371,6 +371,14 @@ req, res, null, forbiddenWithRemoteInfo(req))))
                     .hasAnyRole(MANAGER.getRole(), TEACHER.getRole(), STUDENT.getRole())
                     .requestMatchers(PUT, "/events/*/participants")
                     .hasAnyRole(MANAGER.getRole(), TEACHER.getRole())
+                    .requestMatchers(GET, "/promotions")
+                    .hasAnyRole(MANAGER.getRole(), TEACHER.getRole(), STUDENT.getRole())
+                    .requestMatchers(PUT, "/promotions")
+                    .hasAnyRole(MANAGER.getRole())
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java:411]: hasRole for one Role

File: src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java [hunk 2]
```diff
@@ -371,6 +371,14 @@ req, res, null, forbiddenWithRemoteInfo(req))))
...
+                    .requestMatchers("/promotions/*")
+                    .hasAnyRole(MANAGER.getRole(), TEACHER.getRole(), STUDENT.getRole())
+                    .requestMatchers(PUT, "/promotions/*/groups")
+                    .hasAnyRole(MANAGER.getRole())
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java:415]: hasRole

File: src/main/java/school/hei/haapi/model/Group.java [hunk 3]
```diff
@@ -43,6 +45,10 @@ public class Group implements Serializable {
   @OneToMany(mappedBy = "group", fetch = LAZY)
   private List<GroupFlow> groupFlows;
 
+  @ManyToOne
+  @JoinColumn(name = "promotion_id", referencedColumnName = "id")
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Group.java:52]: referencedColumnName

File: src/main/java/school/hei/haapi/service/GroupService.java [hunk 4]
```diff
@@ -127,4 +128,10 @@ public List<Group> getByUserId(String studentId) {
     }
     return groups;
   }
+
+  public void updateGroups(Promotion promotion, String groupId) {
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/GroupService.java:132]: bad naming

File: src/main/java/school/hei/haapi/service/PromotionService.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public Promotion crupdatePromotion(Promotion domain) {
+    Optional<Promotion> optionalPromotion = promotionRepository.findById(domain.getId());
+    if (optionalPromotion.isPresent()) {
+      optionalPromotion.get().setRef(domain.getRef());
+      optionalPromotion.get().setName(domain.getName());
+      return promotionRepository.save(optionalPromotion.get());
+    }
+    return promotionRepository.save(domain);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/PromotionService.java:47]: i do not understand why you can't directly save

File: src/main/java/school/hei/haapi/service/PromotionService.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public Promotion updateGroups(String promotionId, UpdatePromotionSGroup updatePromotionSGroup) {
+    Promotion promotion = getPromotionById(promotionId);
+    for (String groupId : updatePromotionSGroup.getGroupIds()) {
+      switch (updatePromotionSGroup.getType()) {
+        case ADD -> groupService.updateGroups(promotion, groupId);
+        case REMOVE -> groupService.updateGroups(null, groupId);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/PromotionService.java:55]: use two separate methods ( remove from promotion and add to promotion)

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/model/Group.java (hunk 3 above)
Comment (Iloniavo) [reply src/main/java/school/hei/haapi/model/Group.java:52]: I don't understand

**Pull request merged**